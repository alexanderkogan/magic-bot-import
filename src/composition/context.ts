import * as React from 'react'
import { BookApi, BookApiImpl } from '../test-page/api/fetch-book-info'

export interface AppContext {
    books: BookApi
}
export interface AppContextProps {
    context: AppContext
}
export const productionContext: AppContext = {
    books: new BookApiImpl(),
}
export const AppContext: React.Context<AppContext> = React.createContext(productionContext)
