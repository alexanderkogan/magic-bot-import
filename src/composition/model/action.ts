export interface Action<Name> {
    type: Name
}
