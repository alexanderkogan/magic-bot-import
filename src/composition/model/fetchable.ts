export type Fetchable<T> = Empty | Loading | Loaded<T> | Failed

export interface Empty {
    type: 'empty'
}
export interface Loading {
    type: 'loading'
}
export interface Loaded<T> {
    type: 'loaded'
    value: T
}
export interface Failed {
    type: 'failed'
    value: Error
}

export function empty(): Empty {
    return { type: 'empty' }
}
export function loading(): Loading {
    return { type: 'loading' }
}
export function loaded<T>(value: T): Loaded<T> {
    return { type: 'loaded', value }
}
export function failed(value: Error): Failed {
    return { type: 'failed', value }
}
