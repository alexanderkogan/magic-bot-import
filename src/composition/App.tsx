import * as React from 'react'
import { hot } from 'react-hot-loader/root'

import { AppContext } from './context'

import TestPage from '../test-page/TestPage'

import { BrowserRouter as Router, Switch, Route } from 'react-router-dom'
import Tab from '../pattern-library/Nav'
import TabItem from '../pattern-library/NavItem'

export interface AppProps {
    context: AppContext
}

export const App = (props: AppProps) => (
    <AppContext.Provider value={props.context}>
        <Router>
            <div>
                <Tab>
                    <TabItem to="/">Home</TabItem>
                    <TabItem to="/test">Pattern Library Test</TabItem>
                </Tab>

                <Switch>
                    <Route path="/test">
                        <TestPage />
                    </Route>
                    <Route path="/">lorem ipsum</Route>
                </Switch>
            </div>
        </Router>
    </AppContext.Provider>
)

export default hot(App)
