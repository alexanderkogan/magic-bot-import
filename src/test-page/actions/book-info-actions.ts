import { Action } from '../../composition/model/action'

export type BookInfoAction = BookInfoFetchedAction | BookInfoFetchFailedAction | BookInfoFetchingAction

export type BookInfoFetchingAction = Action<'BOOK_INFO_FETCHING'>
export type BookInfoFetchedAction = Action<'BOOK_INFO_FETCHED'> & { value: string }
export type BookInfoFetchFailedAction = Action<'BOOK_INFO_FETCH_FAILED'> & { error: Error }

export function fetching(): BookInfoFetchingAction {
    return { type: 'BOOK_INFO_FETCHING' }
}
export function fetched(value: string): BookInfoFetchedAction {
    return { type: 'BOOK_INFO_FETCHED', value }
}
export function failed(error: Error): BookInfoFetchFailedAction {
    return { type: 'BOOK_INFO_FETCH_FAILED', error }
}
