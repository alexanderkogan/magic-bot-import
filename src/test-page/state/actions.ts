import { Action } from '../../composition/model/action'
import { SelectableValue } from '.'

export type TestItemAction = PopTestItemAction | PushTestItemAction

export type PopTestItemAction = Action<'TEST_ITEM_POP'>
export type PushTestItemAction = Action<'TEST_ITEM_PUSH'> & { value: SelectableValue }

export function pushItem(value: SelectableValue): PushTestItemAction {
    return { type: 'TEST_ITEM_PUSH', value }
}

export function popItem(): PopTestItemAction {
    return { type: 'TEST_ITEM_POP' }
}
