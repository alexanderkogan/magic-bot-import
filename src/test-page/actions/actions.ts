import { TestItemAction } from './test-item-actions'
import { BookInfoAction } from './book-info-actions'

export type TestPageAction = TestItemAction | BookInfoAction
