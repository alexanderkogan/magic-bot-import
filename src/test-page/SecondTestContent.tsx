import * as React from 'react'
import { Dispatch } from 'redux'
import { connect } from 'react-redux'

import * as BookInfoActions from './actions/book-info-actions'

import Content from '../pattern-library/Content'
import { RootState, extractTestPage } from '../composition/store'
import { TestPageState } from './state'
import { AsyncReader } from '../async-reader'
import { AppContext, AppContextProps } from '../composition/context'
import { Fetchable } from '../composition/model/fetchable'

export interface SecondTestContentProps {
    load: (isb: string) => AsyncReader<AppContext, void>
    bookInfo: Fetchable<string>
}

export const mapStateToProps = (state: TestPageState) => {
    return { bookInfo: state.bookInfo }
}

export const dispatchToProps = (dispatch: Dispatch) => ({
    load: (isbn: string) => async (context: AppContext) => {
        dispatch(BookInfoActions.fetching())
        try {
            const info = await context.books.fetchBookInfo(isbn)
            if (info !== undefined) {
                dispatch(BookInfoActions.fetched(info.description))
            }
        } catch (e) {
            dispatch(BookInfoActions.failed(e))
        }
    },
})

export class Component extends React.Component<SecondTestContentProps & AppContextProps> {
    public componentDidMount() {
        const harryPotter = '0747532699'
        this.props.load(harryPotter)(this.props.context)
    }

    public render() {
        return <Content>{JSON.stringify(this.props.bookInfo)}</Content>
    }
}

export const componentWithContext = (props: SecondTestContentProps) => {
    return <AppContext.Consumer>{context => <Component context={context} {...props}></Component>}</AppContext.Consumer>
}

export default connect(
    (state: RootState) => mapStateToProps(extractTestPage(state)),
    (dispatch: Dispatch) => dispatchToProps(dispatch)
)(componentWithContext)
