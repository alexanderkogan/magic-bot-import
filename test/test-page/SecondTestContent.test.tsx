import * as React from 'react'

import { Component as SecondTestContent, mapStateToProps, dispatchToProps } from '../../src/test-page/SecondTestContent'
import { mount, mockReturnValue } from '../test-utils'
import { AppContext } from '../../src/composition/context'
import { mockedContext } from '../mocks'
import { empty, loaded, failed } from '../../src/composition/model/fetchable'

import * as BookInfoActions from '../../src/test-page/actions/book-info-actions'

describe('SecondTestContent', () => {
    let context: AppContext
    beforeEach(() => {
        context = mockedContext()
    })

    describe('state mapping', () => {
        it('maps empty state to props', () => {
            const props = mapStateToProps({ valuesSelected: [], bookInfo: empty() })
            expect(props).toEqual({ bookInfo: empty() })
        })
        it('maps filled state to props', () => {
            const props = mapStateToProps({ valuesSelected: [], bookInfo: loaded('some info') })
            expect(props).toEqual({ bookInfo: loaded('some info') })
        })
    })

    describe('dispatch mapping', () => {
        it('dispatches load action', async () => {
            const dispatch = jest.fn()
            mockReturnValue(context.books.fetchBookInfo, Promise.resolve({ description: 'some book description' }))
            await dispatchToProps(dispatch).load('1234')(context)
            expect(dispatch).toHaveBeenCalledWith(BookInfoActions.fetching())
            expect(dispatch).toHaveBeenCalledWith(BookInfoActions.fetched('some book description'))
        })

        it('handles failure of load action', async () => {
            const dispatch = jest.fn()
            mockReturnValue(context.books.fetchBookInfo, Promise.reject(new Error('some error')))
            await dispatchToProps(dispatch).load('1234')(context)
            expect(dispatch).toHaveBeenCalledWith(BookInfoActions.fetching())
            expect(dispatch).toHaveBeenCalledWith(BookInfoActions.failed(new Error('some error')))
        })
    })

    describe('component', () => {
        it('renders', () => {
            const element = mount(<SecondTestContent context={context} load={() => jest.fn()} bookInfo={loaded('some book info')} />)
            expect(element).toMatchSnapshot()
        })

        it('loads on mount', () => {
            const reader = jest.fn()
            mount(<SecondTestContent context={context} load={() => reader} bookInfo={empty()} />)
            expect(reader).toHaveBeenCalledWith(context)
        })
    })
})
