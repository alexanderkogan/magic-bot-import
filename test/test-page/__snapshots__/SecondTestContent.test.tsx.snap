// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecondTestContent component renders 1`] = `
<Component
  bookInfo={
    Object {
      "type": "loaded",
      "value": "some book info",
    }
  }
  context={
    Object {
      "books": Object {
        "fetchBookInfo": [MockFunction],
      },
    }
  }
  load={[Function]}
>
  <Component>
    <div>
      {"type":"loaded","value":"some book info"}
    </div>
  </Component>
</Component>
`;
